design/20210209.certificates.k8s.io-adoption.md

```diff
@@ -0,0 +1,292 @@
+against the `UserInfo` fields on the `CertificateSigningRequest`. The required
+role must be of `certificates.k8s.io/CertificateSigningRequest` with the verb
+`create` within the target namespace.
+
+Since `Certificates` are also namespaced resources, cert-manager validates that
+created `Certificate` resources may only reference `Issuers` in the same
+namespace, or `ClusterIssuers`. Any resulting `CertificateSigningRequests` will
+reference `Issuers` in the same namespace.
+
+### Duration Field
+
+`CertificateSigningRequests` do not include a `duration` field. To have parity
+with the `CertificateRequest` resource, the duration field will be moved to an
+annotation as `cert-manager.io/duration` which includes a [Go time duration
+string](https://golang.org/pkg/time/#Duration.String). This annotation may not
+be changed after creation.
+
+### CA Field
+
+`CertificateSigningRequests` do not include a `ca` field. To have parity with
+the `CertificateRequest` resource, the `ca` field will be moved to an annotation
+as `cert-manager.io/ca`. This annotation value contains the base 64 PEM encoded
+CA certificate (if one was returned by the signer). This annotation may not be
+changed after creation.
+
+### External Issuers
+
+All current [external
+issuers](https://cert-manager.io/docs/configuration/external/) are built for the
+`CertificateRequest` resource. The project should continue to support this
+resource indefinitely, and provide an example project for creating external
+signers if and when the migration to `CertificateSigningRequests` occurs.
+
+
+---
+
+
+## Proposal
+
+cert-manager will support both `CertificateRequest` and
+`CertificateSigningRequest` types until, and if, the project completely migrates
+to the `CertificateSigningRequest` resource. The `Certificates` controller will
+continue to create `CertificateRequest` resources, unless a flag is changed on
+the cert-manager controller. This default can be switched in future.
+
+The cert-manager controller will not enable the `CertificateSigningRequest`
+controllers unless the resource is available in the running cluster. This can be
+checked via the [discovery endpoint of the API
+server](https://github.com/kubernetes/client-go/blob/7279fc64d8478bd5f9d38122143b5e6294f06e75/discovery/discovery_client.go#L55).
+If the resource is not available, the `CertificateSigningRequest` controllers
+should gracefully never start. If the Kubernetes API server were to be upgraded
+to a version that does support this resource, cert-manager will need to be
+restarted to make use of these controllers. This is acceptable- worker nodes are
+typically always restarted during a cluster upgrade.
+
+
+Instead of the concept of an `IssuerRef` for `CertificateRequests`,
+`CertificateSigningRequests` have the concept of a `SignerName`. Since
+`CertificateSigningRequests` are cluster scoped resources, the signer name must
```

Comment:
By _must_, do you mean that this will be enforced (e.g., using a webhook)?

Reply:
Must, as in, it would be referencing a cluster issuer if the namespace is not present.